new-log-viewer/src/utils/data.ts

```diff
@@ -1,6 +1,83 @@
 import {Nullable} from "../typings/common";
 
 
+/**
+ * Checks if `target` is bounded by the first and last value in a sorted array of numbers.
+ *
+ * @param data An array sorted in ascending order.
+ * @param target
+ * @return Whether `target` is within the bounds of the array's values.
+ */
+const isWithinBounds = (data: number[], target: number): boolean => {
+    const {length} = data;
+    if (0 === length) {
+        return false;
+    }
+
+    return (target >= (data[0] as number)) && (target <= (data[length - 1] as number));
+};
+
+/**
+ * Performs binary search to find the smallest index `i` in the range `[0, length)` where
+ * `predicate` is true. `predicate` should be false for some prefix of the input range and true
+ * for the remainder.
+ *
+ * @param length The length of the range to search.
+ * @param predicate A function that takes an index and returns `true` or `false`.
+ * @return The smallest index where `predicate(i)` is true, or `length` if no such index exists.
+ * @example
+ * const arr = [1, 3, 5, 7, 10, 15, 20];
+ * const result = binarySearch(arr.length, (i) => arr[i] >= 10);
+ * console.log(result); // Output: 4 (since arr[4] is 10).
+ */
+const binarySearch = (length: number, predicate: (index: number) => boolean): number => {
+    // Generic implementation based on Go standard library implementation.
+    // Reference: https://pkg.go.dev/sort#Search
+    let i = 0;
+    let j = length;
+    while (i < j) {
+        const mid = Math.floor((i + j) / 2);
+        if (false === predicate(mid)) {
+            i = mid + 1;
+        } else {
+            j = mid;
+        }
+    }
+
+    return i;
+};
+
+/**
+ * Finds the largest index `i` in a sorted array `data` such that `data[i] <= target`. Uses binary
+ * search for efficiency.
```

Comment:
```suggestion
 * Finds the index `i` in a sorted array `data` such that `data[i]` is the "first nearest" element
 * to `target`, where "first nearest" is defined as follows:
 * - If `target` is in or greater than the range of `data`'s elements, then `i` is the largest index
 *   such that `data[i] <= target`.
 * - Otherwise, `target` is less than the range of `data`'s elements, so `i` is 0 since `data[0]` is
 *   the array element closest to `target`.
```